src/uu/pwd/src/pwd.rs

```diff
@@ -10,28 +10,115 @@ extern crate uucore;
 
 use clap::{crate_version, App, Arg};
 use std::env;
-use std::io;
+use std::io::{self, Write};
 use std::path::{Path, PathBuf};
 
-use uucore::error::{FromIo, UResult, USimpleError};
+use uucore::error::{FromIo, UResult};
 
 static ABOUT: &str = "Display the full filename of the current working directory.";
 static OPT_LOGICAL: &str = "logical";
 static OPT_PHYSICAL: &str = "physical";
 
-pub fn absolute_path(path: &Path) -> io::Result<PathBuf> {
-    let path_buf = path.canonicalize()?;
+fn physical_path() -> io::Result<PathBuf> {
+    // std::env::current_dir() is a thin wrapper around libc's getcwd().
 
+    // On Unix, getcwd() must return the physical path:
+    // https://pubs.opengroup.org/onlinepubs/9699919799/functions/getcwd.html
+    #[cfg(unix)]
+    {
+        env::current_dir()
+    }
+
+    // On Windows we have to resolve it.
+    // On other systems we also resolve it, just in case.
+    #[cfg(not(unix))]
+    {
+        env::current_dir().and_then(|path| path.canonicalize())
+    }
+}
+
+fn logical_path() -> io::Result<PathBuf> {
+    // getcwd() on Windows seems to include symlinks, so this is easy.
     #[cfg(windows)]
-    let path_buf = Path::new(
-        path_buf
-            .as_path()
-            .to_string_lossy()
-            .trim_start_matches(r"\\?\"),
-    )
-    .to_path_buf();
-
-    Ok(path_buf)
+    {
+        env::current_dir()
+    }
+
+    // If we're not on Windows we do things Unix-style.
+    //
+    // Typical Unix-like kernels don't actually keep track of the logical working
+    // directory. They know the precise directory a process is in, and the getcwd()
+    // syscall reconstructs a path from that.
+    //
+    // The logical working directory is maintained by the shell, in the $PWD
+    // environment variable. So we check carefully if that variable looks
+    // reasonable, and if not then we fall back to the physical path.
+    //
+    // POSIX: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pwd.html
+    #[cfg(not(windows))]
+    {
+        fn looks_reasonable(path: &Path) -> bool {
+            // First, check if it's an absolute path.
+            if !path.has_root() {
+                return false;
+            }
+
+            // Then, make sure there are no . or .. components.
+            // Path::components() isn't useful here, it normalizes those out.
+
+            // to_string_lossy() may allocate, but that's fine, we call this
+            // only once per run. It may also lose information, but not any
+            // information that we need for this check.
+            if path
+                .to_string_lossy()
+                .split(std::path::is_separator)
+                .any(|piece| piece == "." || piece == "..")
+            {
+                return false;
+            }
+
+            // Finally, check if it matches the directory we're in.
+            #[cfg(unix)]
```

Comment:
FYI: I'm adding a common way to compare files in https://github.com/uutils/coreutils/pull/2610/commits/80b804360b05465a900302783480ba07eeda4f1d (#2610).